src/stactools/noaa_hrrr/stac.py

```diff
@@ -0,0 +1,562 @@
+            REFERENCE_TIME: Dimension(
+                properties={
+                    "type": DimensionType.TEMPORAL,
+                    "extent": [
+                        cloud_provider_config.start_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+                        None,
+                    ],
+                    "step": "PT1H" if region == Region.conus else "PT3H",
+                }
+            ),
+            VALID_TIME: Dimension(
+                properties={
+                    "type": DimensionType.TEMPORAL,
+                    "extent": [
+                        cloud_provider_config.start_date.strftime("%Y-%m-%dT%H:%M:%SZ"),
+                        None,
+                    ],
+                    "step": "PT15M" if product == Product.subh else "PT1H",
+                }
+            ),
+            # could be a z spatial dimension but the units are not consistent
+            LEVEL: Dimension(
+                properties={
+                    "type": "atmospheric level",
+                    "description": (
+                        "The atmospheric level for which the forecast is applicable, "
+                        "e.g. surface, top of atmosphere, 100 m above ground, etc."
+                    ),
+                    "values": list(sorted(set(inventory_df[LEVEL].unique()))),
+                }
+            ),
+            FORECAST_VALID: Dimension(
+                properties={
+                    "type": DimensionType.TEMPORAL,
+                    "description": (
+                        "The time horizon for which the forecast is applicable."
+                    ),
+                    "values": list(sorted(set(inventory_df[FORECAST_VALID].unique()))),
+                }
+            ),
+        },
+        variables={
+            variable: Variable(
+                properties=dict(
+                    dimensions=[
+                        "x",
+                        "y",
+                        REFERENCE_TIME,
+                        VALID_TIME,
+                        LEVEL,
+                        FORECAST_VALID,
+                    ],
+                    type=VariableType.DATA,
+                    description=description,
+                    unit=unit,
+                    # experimental new field for defining the specific values of each
+                    # domain where this variable has data
+                    dimension_domains={
+                        LEVEL: list(group[LEVEL].unique()),
+                        FORECAST_VALID: list(group[FORECAST_VALID].unique()),
+                    },
+                )
+            )
+            for (variable, description, unit), group in variable_df.groupby(
+                level=[VARIABLE, DESCRIPTION, UNIT]
+            )
+        },
+    )
+    return collection
+
+
+def create_item(
+    region: Region,
+    product: Product,
+    cloud_provider: CloudProvider,
+    reference_datetime: datetime,
+    forecast_hour: int,
+) -> Item:
```

Comment:
@TomAugspurger - I can make it possible to just create an item from a GRIB href here. I think you also suggested that it would be useful to be able to pass in the the index dataframe rather than reading it from the blob storage url in the `create_item` function,

Reply:
I think my ideal set of APIs is something like

```python
def create_item(href: str, product: Product, reference_datetime: datetime, forecast_hour: int) -> Item: ...
``` 

Where `href` is the path to either the grib2 file or the index file (the grib2 file would probably more consistent with other stactools packages).

And to make a (hopefully) IO-free version, something like

```python
def create_item_from_inventory(inventory: pandas.DataFrame, href: str, product: Product, reference_datetime: datetime, forecast_hour: int) -> Item: ...
```

Then ideally `create_item` just reads the inventory file and calls `create_item_from_inventory`.

I haven't really looked at how my `href` would interact with the `reference_datetime` and `product` arguments, but maybe they're duplicative of each other?

---

One other note on how our data pipelines and stactools packages interact: our pipelines will typically get a list of HREFs / prefixes that we want to process, and then we'll call the stactools package's `create_item` on each element of that list. Roughly, I'm hoping for something like

```python
# in planetary-computer-tasks
hrefs = list_hrefs(...)  # list of hrefs in the same collection, one per item. e.g. 'https://noaahrrr.blob.core.windows.net/hrrr/hrrr.20210321/conus/hrrr.t00z.wrfnatf00.grib2'
for href in hrefs:
    inventory = download_inventory(href + ".idx")
    item = stac.create_item_from_inventory(inventory, href, ...)

```

Reply:
That all makes sense to me - I will make those API changes.

> I haven't really looked at how my href would interact with the reference_datetime and product arguments, but maybe they're duplicative of each other?

The `region`, `product`, `reference_datetime`, and `forecast_hour` parameters are all represented in the hrefs, so including all of them would be redundant. I set it up with the parameters so that a user could generate items without crawling the whole storage container. I could refactor the function a bit to optionally take either the parameter args or an href, or I could add a convenience function to parse an href into a set of args:
```python
# in planetary-computer-tasks
hrefs = list_hrefs(...)  # list of hrefs in the same collection, one per item. e.g. 'https://noaahrrr.blob.core.windows.net/hrrr/hrrr.20210321/conus/hrrr.t00z.wrfnatf00.grib2'
for href in hrefs:
    inventory = download_inventory(href + ".idx")
    item = stac.create_item_from_inventory(inventory, **parse_href(href))
```

Reply:
> I set it up with the parameters so that a user could generate items without crawling the whole storage container.

Ah, I guess since these are model output, we (in theory) know exactly what files should be there. This does seem useful.

I think for dynamically updating pipelines, we'd also want the option to start from an HREF (liked you'd get in a Cluod Event as new files are written). A simple helper to parse a `Region`, `Product`, etc. from an HREF would work for us I think.

Reply:
This is all set, docs/example.ipynb shows how to use the variations on `create_item`.